File: DataStructure/1570.dot-product-of-two-sparse-vectors.ts
// @leet start
class SparseVector {
  nonZero = new Map<number, number>();
  constructor(nums: number[]) {
    for (const [idx, v] of nums.entries()) {
      if (v !== 0) {
        this.nonZero.set(idx, v);
      }
    }
  }

  // Return the dotProduct of two sparse vectors
  dotProduct(vec: SparseVector): number {
    let res = 0;
    for (const [k, v] of this.nonZero) {
      res += v * (vec.nonZero.get(k) ?? 0);
    }
    return res;
  }
}

/**
 * Your SparseVector object will be instantiated and called as such:
 * var v1 = new SparseVector(nums1)
 * var v2 = new SparseVector(nums1)
 * var ans = v1.dotProduct(v2)
 */
// @leet end

File: DataStructure/1570.dot-product-of-two-sparse-vectors.go
package leetcode

// @leet start
type SparseVector struct {
	nonZero map[int]int
}

func Constructor(nums []int) SparseVector {
	nonZero := map[int]int{}
	for i, v := range nums {
		if v != 0 {
			nonZero[i] = v
		}
	}
	return SparseVector{
		nonZero: nonZero,
	}
}

// Return the dotProduct of two sparse vectors
func (this *SparseVector) dotProduct(vec SparseVector) int {
	result := 0
	for i, v := range this.nonZero {
		result += v * vec.nonZero[i]
	}
	return result
}

/**
 * Your SparseVector object will be instantiated and called as such:
 * v1 := Constructor(nums1);
 * v2 := Constructor(nums2);
 * ans := v1.dotProduct(v2);
 */
// @leet end

